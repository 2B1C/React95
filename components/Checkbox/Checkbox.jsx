import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import check from '../../assets/pattern/check.svg';

const Field = styled.input`
  display: none;
`;

const Label = styled.label`
  position: relative;
  padding-left: 26px;
  height: 15px;
  display: inline-block;
  line-height: 15px;
  background-repeat: no-repeat;
  background-position: 0 0;
  font-size: 15px;
  vertical-align: middle;

  &:before {
    content: '';
    position: absolute;
    left: 0;
    display: inline-block;
    height: 15px;
    width: 15px;

    border-left: 1px solid #868a8e;
    border-top: 1px solid #868a8e;

    box-shadow: inset -1px -1px 0 0 #c3c7cb, inset 1px 1px 0 0 #000000,
      0.5px 0.5px 0 0.5px #ffffff;
    background-color: white;
  }

  ${({ checked }) =>
    checked &&
    `
  &:before {
    background-image: url('${check}');
    background-repeat: no-repeat;
    background-position: center center;
    background-size: 10px 10px;
  }`};
`;

class Checkbox extends React.Component {
  render() {
    const { value, style, label, checked } = this.props;

    return (
      <Label {...style} checked={checked}>
        <Field type="checkbox" value={value} />
        {label}
      </Label>
    );
  }
}

Checkbox.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  checked: PropTypes.bool,
};

Checkbox.defaultProps = {
  value: 'checked',
  label: 'Checkbox',
  checked: false,
};

export default Checkbox;
