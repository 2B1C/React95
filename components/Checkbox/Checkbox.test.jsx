import React from 'react';
import Checkbox from './Checkbox';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

describe('<Checkbox />', () => {
  describe('Snapshots', () => {
    it('should match snapshot without props', () => {
      expect(renderer.create(<Checkbox />).toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with checked prop', () => {
      expect(renderer.create(<Checkbox checked />).toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with disabled prop', () => {
      expect(renderer.create(<Checkbox disabled />).toJSON()).toMatchSnapshot();
    });

    it('should match snapshot with checked and disabled props', () => {
      expect(
        renderer.create(<Checkbox checked disabled />).toJSON()
      ).toMatchSnapshot();
    });
  });

  describe('Checkbox label', () => {
    it('should display label prop as Checkbox label', () => {
      const label = 'label text';
      const ShallowCheckbox = shallow(<Checkbox label={label} />);

      expect(ShallowCheckbox.text()).toBe(label);
    });

    it('should display children as Checkbox label', () => {
      const label = 'label text';
      const ShallowCheckbox = shallow(<Checkbox>{label}</Checkbox>);

      expect(ShallowCheckbox.text()).toBe(label);
    });
  });

  describe('onChange prop', () => {
    it('should call onChange function when Checkbox is changed', () => {
      const onChangeMock = jest.fn();
      const ShallowCheckbox = shallow(
        <Checkbox onChange={onChangeMock} label="text" />
      );

      ShallowCheckbox.find('Field').simulate('change');

      expect(onChangeMock).toHaveBeenCalled();
    });
  });
});
