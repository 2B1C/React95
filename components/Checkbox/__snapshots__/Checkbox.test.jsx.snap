// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> Snapshots should match snapshot with checked and disabled props 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 1px;
}

.c1 {
  height: 12px;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 12px;
  margin: 0;
}

.c1:focus + .c2,
.c1:active + .c2 {
  border: 1px dotted;
  padding: 0;
}

.c0 {
  position: relative;
  padding-left: 20px;
  height: 15px;
  line-height: 1.5;
  display: inline-block;
  margin: 4px 0;
  color: #868686;
  text-shadow: 0.5px 0.5px #d2d2d2;
}

.c0:before {
  content: '';
  position: absolute;
  left: 0;
  display: inline-block;
  height: 12px;
  width: 12px;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
}

.c0:before {
  background-image: url('test-file-stub'),url('test-file-stub');
  background-repeat: no-repeat,repeat;
  background-position: center center;
  background-size: 7px 7px,1.9px 1.9px;
}

.c0:before {
  background-color: #c0c0c0;
}

<label
  checked={true}
  className="c0"
  disabled={true}
>
  <input
    checked={true}
    className="c1"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
    value="checked"
  />
  <span
    className="c2 c3"
  >
    Checkbox
  </span>
</label>
`;

exports[`<Checkbox /> Snapshots should match snapshot with checked prop 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 1px;
}

.c1 {
  height: 12px;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 12px;
  margin: 0;
}

.c1:focus + .c2,
.c1:active + .c2 {
  border: 1px dotted;
  padding: 0;
}

.c0 {
  position: relative;
  padding-left: 20px;
  height: 15px;
  line-height: 1.5;
  display: inline-block;
  margin: 4px 0;
}

.c0:before {
  content: '';
  position: absolute;
  left: 0;
  display: inline-block;
  height: 12px;
  width: 12px;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
}

.c0:before {
  background-image: url('test-file-stub');
  background-repeat: no-repeat;
  background-position: center center;
  background-size: 7px 7px;
}

<label
  checked={true}
  className="c0"
  disabled={false}
>
  <input
    checked={true}
    className="c1"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
    value="checked"
  />
  <span
    className="c2 c3"
  >
    Checkbox
  </span>
</label>
`;

exports[`<Checkbox /> Snapshots should match snapshot with disabled prop 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 1px;
}

.c1 {
  height: 12px;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 12px;
  margin: 0;
}

.c1:focus + .c2,
.c1:active + .c2 {
  border: 1px dotted;
  padding: 0;
}

.c0 {
  position: relative;
  padding-left: 20px;
  height: 15px;
  line-height: 1.5;
  display: inline-block;
  margin: 4px 0;
  color: #868686;
  text-shadow: 0.5px 0.5px #d2d2d2;
}

.c0:before {
  content: '';
  position: absolute;
  left: 0;
  display: inline-block;
  height: 12px;
  width: 12px;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
}

.c0:before {
  background-color: #c0c0c0;
}

<label
  checked={false}
  className="c0"
  disabled={true}
>
  <input
    checked={false}
    className="c1"
    disabled={true}
    onChange={[Function]}
    type="checkbox"
    value="checked"
  />
  <span
    className="c2 c3"
  >
    Checkbox
  </span>
</label>
`;

exports[`<Checkbox /> Snapshots should match snapshot without props 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 1px;
}

.c1 {
  height: 12px;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  width: 12px;
  margin: 0;
}

.c1:focus + .c2,
.c1:active + .c2 {
  border: 1px dotted;
  padding: 0;
}

.c0 {
  position: relative;
  padding-left: 20px;
  height: 15px;
  line-height: 1.5;
  display: inline-block;
  margin: 4px 0;
}

.c0:before {
  content: '';
  position: absolute;
  left: 0;
  display: inline-block;
  height: 12px;
  width: 12px;
  border-left: 1px solid #868a8e;
  border-top: 1px solid #868a8e;
  box-shadow: inset -1px -1px 0 0 #c3c7cb,inset 1px 1px 0 0 #000000,0.5px 0.5px 0 0.5px #ffffff;
  background-color: white;
}

<label
  checked={false}
  className="c0"
  disabled={false}
>
  <input
    checked={false}
    className="c1"
    disabled={false}
    onChange={[Function]}
    type="checkbox"
    value="checked"
  />
  <span
    className="c2 c3"
  >
    Checkbox
  </span>
</label>
`;
